cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,650 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/types.hpp>
+#include <strings/utilities.cuh>
+#include <strings/convert/utilities.cuh>
+
+#include <map>
+#include <vector>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+// used to index values in a timeparts array
+enum duration_parse_component {
+  DU_DAY       = 0,
+  DU_HOUR      = 1,
+  DU_MINUTE    = 2,
+  DU_SECOND    = 3,
+  DU_SUBSECOND = 4,
+  DU_ARRAYSIZE = 5
+};
+
+enum class format_char_type : int8_t {
+  literal,   // literal char type passed through
+  specifier  // duration format specifier
+};
+
+/**
+ * @brief Represents a format specifier or literal from a duration format string.
+ *
+ * Created by the format_compiler when parsing a format string.
+ */
+struct alignas(4) format_item {
+  format_char_type item_type;  // specifier or literal indicator
+  char value;                  // specifier or literal value
+  int8_t length;               // item length in bytes
+
+  static format_item new_specifier(char format_char, int8_t length)
+  {
+    return format_item{format_char_type::specifier, format_char, length};
+  }
+  static format_item new_delimiter(char literal)
+  {
+    return format_item{format_char_type::literal, literal, 1};
+  }
+};
+
+/**
+ * @brief The format_compiler parses a duration format string into a vector of
+ * format_items.
+ *
+ * The vector of format_items are used when parsing a string into duration
+ * components and when formatting a string from duration components.
+ */
+struct format_compiler {
+  std::string format;
+  rmm::device_vector<format_item> d_items;
+
+  std::map<char, int8_t> specifier_lengths = {{'d', -1},
```

Comment:
```suggestion
  static std::map<char, int8_t> const specifier_lengths = {{'d', -1},
```

Reply:
Multi-threaded access will not be safe for a static member.
just wondering, should libcudf calls always be single threaded?

Reply:
Good point! Use a function local static (i.e., "singleton pattern) and then it will be thread safe: https://stackoverflow.com/questions/1661529/is-meyers-implementation-of-the-singleton-pattern-thread-safe

Reply:
Updated to function local static.